src/Castle.Windsor/MicroKernel/Registration/BasedOnDescriptor.cs

```diff
@@ -535,11 +536,14 @@ private static bool IsBasedOnGenericInterface(Type type, Type basedOn, out Type[
 			var types = new List<Type>(4);
 			foreach (var @interface in type.GetInterfaces())
 			{
-				if (@interface.IsGenericType &&
+				if (@interface.GetTypeInfo().IsGenericType &&
 				    @interface.GetGenericTypeDefinition() == basedOn)
 				{
-					if (@interface.ReflectedType == null &&
-					    @interface.ContainsGenericParameters)
+					if (@interface.GetTypeInfo().ContainsGenericParameters
+#if FEATURE_REFLECTION_REFLECTEDTYPE
+						&& @interface.ReflectedType == null
+#endif
```

Comment:
Because in both these places `ReflectedType` is just used to check for null (and not actually doing anything with the ReflectedType), I would say it doesn't make a difference if it's the base or derived type returned and therefore replacing it with `DeclaringType` is safe.